desc/vmec_utils.py

```diff
@@ -468,3 +482,451 @@ def vmec_boundary_subspace(eq, RBC=None, ZBS=None, RBS=None, ZBC=None):  # noqa:
+    verbose: int
+        Level of output (Default = 1).
+        * 0: no output
+        * 1: status of quantities computed
+        * 2: as above plus timing information
+
+    Returns
+    -------
+    None
+
+    """
+    timer = Timer()
+    timer.start("Total time")
+
+    Psi = eq.Psi
+    NFP = eq.NFP
+    if M_booz is None:
+        M_booz = 2 * eq.M
+    if N_booz is None:
+        N_booz = 2 * eq.N
+
+    # calculations are done on a "half-grid" of size surfs-1
+    # which starts at s =  psi = 0.5/(surfs-1)
+    # and increments by 1 / (surfs-1)
+    # until it reaches 1-0.5/(surfs-1)
+    # VMEC radial coordinate: s = rho^2 = Psi / Psi(LCFS)
+    s_full = np.linspace(0, 1, surfs)
+    hs = 1 / (surfs - 1)
+    s_half = s_full[0:-1] + hs / 2
+    r_full = np.sqrt(s_full)
+    r_half = np.sqrt(s_half)
+
+    grid = LinearGrid(M=2 * M_booz, N=2 * N_booz, NFP=eq.NFP, rho=r_half, sym=False)
+
+    transforms = get_transforms(
+        "|B|_mn_B",
+        obj=eq,
+        grid=grid,
+        M_booz=M_booz - 1,
+        N_booz=N_booz,
+    )
+    basis = transforms["B"].basis
+
+    matrix, modes = ptolemy_linear_transform(basis.modes)
+    # if sym is False, then the number of modes is double what each individual mode
+    # array should be, since it has both sin(mt - nz) and cos(mt-nz) modes in it
+    # while num_modes is the number of modes for a single sin or cos series
+    num_modes = modes.shape[0] if eq.sym else int((modes.shape[0] + 1) / 2)
+
+    if eq.sym:  # need a separate sin basis for Z and nu
+        transforms_sin = get_transforms(
+            "|B|_mn_B",
+            obj=eq,
+            grid=grid,
+            M_booz=M_booz - 1,
+            N_booz=N_booz,
+            sym="sin",
+        )
+        basis_sin = transforms_sin["B"].basis
+        matrix_sin, modes_sin = ptolemy_linear_transform(basis_sin.modes)
+
+    else:
+        matrix_sin = matrix
+        modes_sin = modes
+        transforms_sin = transforms
+        basis_sin = basis
+
+    timer.start("Boozer Transform")
+
+    # precompute the needed data for the boozer surface computations
+    keys = [
+        "|B|",
+        "R",
+        "Z",
+        "sqrt(g)",
+        "rho",
+        "psi_r",
+        "lambda",
+        "B_zeta",
+        "B_theta",
+        "G",
+        "I",
+        "lambda_t",
+        "lambda_z",
+        # need to precompute these so that when eq.sym==True,
+        # the sin-symmetric coefficients have the correct values,
+        # since these quantities below rely on cos-symmetric transforms
+        "w_Boozer",
+        "nu",
+        "sqrt(g)_Boozer_DESC",
+    ]
+    data_vol = eq.compute(keys, grid=grid)
+
+    data_keys = ["|B|_mn_B", "R_mn_B", "sqrt(g)_B_mn", "psi_r"]
+    data_keys = data_keys + ["Z_mn_B", "nu_mn"] if not eq.sym else data_keys
+    data_keys_sin = ["Z_mn_B", "nu_mn"]
+
+    data = eq.compute(data_keys, grid=grid, data=data_vol.copy(), transforms=transforms)
```

Comment:
does this have to be a separate step? didn't we just compute a bunch of stuff on the same grid?

Reply:
it's separate at least from the below one because if stellarator symmetry is assumed, the sin-symmetric quantites Z_mn_b and nu_mn cannot be computed with the same transforms as the cos-symmetric quantities. So I separate out these steps so that I can control that. 

I guess this and the above can be combined and then I can just compute the sin stuff separately if needed though.